deps/rabbitmq_auth_backend_oauth2/test/jwks_SUITE.erl

```diff
@@ -0,0 +1,389 @@
+    rabbit_ct_helpers:testcase_started(Config, Testcase),
+    Config;
+
+init_per_testcase(Testcase, Config) when Testcase =:= test_failed_token_refresh_case1 orelse
+                                         Testcase =:= test_failed_token_refresh_case2 ->
+    rabbit_ct_broker_helpers:add_vhost(Config, <<"vhost4">>),
+    rabbit_ct_helpers:testcase_started(Config, Testcase),
+    Config;
+
+init_per_testcase(Testcase, Config) when Testcase =:= test_successful_connection_with_complex_claim_as_a_map orelse
+                                         Testcase =:= test_successful_connection_with_complex_claim_as_a_list orelse
+                                         Testcase =:= test_successful_connection_with_complex_claim_as_a_binary ->
+  ok = rabbit_ct_broker_helpers:rpc(Config, 0, application, set_env,
+        [rabbitmq_auth_backend_oauth2, extra_scopes_source, ?EXTRA_SCOPES_SOURCE]),
+  rabbit_ct_helpers:testcase_started(Config, Testcase),
+  Config;
+
+init_per_testcase(Testcase, Config) ->
+    rabbit_ct_helpers:testcase_started(Config, Testcase),
+    Config.
+
+end_per_testcase(Testcase, Config) when Testcase =:= test_failed_token_refresh_case1 orelse
+                                        Testcase =:= test_failed_token_refresh_case2 ->
+    rabbit_ct_broker_helpers:delete_vhost(Config, <<"vhost4">>),
+    rabbit_ct_helpers:testcase_started(Config, Testcase),
+    Config;
+
+end_per_testcase(Testcase, Config) when Testcase =:= test_successful_connection_with_complex_claim_as_a_map orelse
+                                        Testcase =:= test_successful_connection_with_complex_claim_as_a_list orelse
+                                        Testcase =:= test_successful_connection_with_complex_claim_as_a_binary ->
+  rabbit_ct_broker_helpers:delete_vhost(Config, <<"vhost1">>),
+  ok = rabbit_ct_broker_helpers:rpc(Config, 0, application, set_env,
+    [rabbitmq_auth_backend_oauth2, extra_scopes_source, undefined]),
+  rabbit_ct_helpers:testcase_started(Config, Testcase),
+  Config;
+
+end_per_testcase(Testcase, Config) ->
+    rabbit_ct_broker_helpers:delete_vhost(Config, <<"vhost1">>),
+    rabbit_ct_helpers:testcase_finished(Config, Testcase),
+    Config.
+
+preconfigure_node(Config) ->
+    ok = rabbit_ct_broker_helpers:rpc(Config, 0, application, set_env,
+                                      [rabbit, auth_backends, [rabbit_auth_backend_oauth2]]),
+    ok = rabbit_ct_broker_helpers:rpc(Config, 0, application, set_env,
+                                      [rabbitmq_auth_backend_oauth2, resource_server_id, ?RESOURCE_SERVER_ID]),
+    Config.
+
+start_jwks_server(Config) ->
+    Jwk   = ?UTIL_MOD:fixture_jwk(),
+    %% Assume we don't have more than 100 ports allocated for tests
+    PortBase = rabbit_ct_broker_helpers:get_node_config(Config, 0, tcp_ports_base),
+    JwksServerPort = PortBase + 100,
+    ok = application:set_env(jwks_http, keys, [Jwk]),
+    {ok, _} = application:ensure_all_started(cowboy),
```

Comment:
This leaks the cowboy application, but that seems like a better option than forcibly shutting down all cowboys at the end, regardless of whether they were already running since before?

Reply:
Starting and not shutting down a Cowboy listener for the duration of this suite sounds acceptable.

Reply:
Alright.